src/libasr/codegen/llvm_utils.cpp

```diff
@@ -2237,6 +2237,59 @@ namespace LFortran {
         shift_end_point_by_one(list);
     }
 
+    llvm::Value* LLVMList::list_concat(llvm::Value* left, llvm::Value* right,
+                    ASR::ttype_t* asr_type, llvm::Module& module) {
+        std::string type_code = ASRUtils::get_type_code(asr_type);
+        int type_size = std::get<1>(typecode2listtype[type_code]);
+        llvm::Type* el_type = std::get<2>(typecode2listtype[type_code]);
+        llvm::Value* new_list = builder->CreateAlloca(llvm_utils->list_api->get_list_type(el_type,
+                                                          type_code, type_size), nullptr);
+        llvm_utils->list_api->list_init(type_code, new_list, module, 0, 0);
+        list_deepcopy(left, new_list, asr_type, module);
+
+        llvm::Value* rlist_end_point = LLVM::CreateLoad(*builder,
+                                        get_pointer_to_current_end_point(right));
+        llvm::Value* zero = llvm::ConstantInt::get(llvm::Type::getInt32Ty(context),
+                                                llvm::APInt(32, 0));
+        // TODO: Should be created outside the user loop and not here.
+        // LLVMList should treat them as data members and create them
+        // only if they are NULL
+        llvm::AllocaInst *pos_ptr = builder->CreateAlloca(
+                                    llvm::Type::getInt32Ty(context), nullptr);
+        LLVM::CreateStore(*builder, zero, pos_ptr);
+        llvm::Value* tmp = nullptr;
+
+        llvm::BasicBlock *loophead = llvm::BasicBlock::Create(context, "loop.head");
+        llvm::BasicBlock *loopbody = llvm::BasicBlock::Create(context, "loop.body");
+        llvm::BasicBlock *loopend = llvm::BasicBlock::Create(context, "loop.end");
+
+        // head
+        llvm_utils->start_new_block(loophead);
+        {
+            llvm::Value *cond = builder->CreateICmpSGT(
+                                        rlist_end_point,
+                                        LLVM::CreateLoad(*builder, pos_ptr));
+            builder->CreateCondBr(cond, loopbody, loopend);
+        }
+
+        // body
+        llvm_utils->start_new_block(loopbody);
+        {
+            tmp = read_item(right, LLVM::CreateLoad(*builder, pos_ptr), false);
+            append(new_list, tmp, asr_type, module);
+            tmp = builder->CreateAdd(
+                        LLVM::CreateLoad(*builder, pos_ptr),
+                        llvm::ConstantInt::get(context, llvm::APInt(32, 1)));
+            LLVM::CreateStore(*builder, tmp, pos_ptr);
+        }
+        builder->CreateBr(loophead);
+
+        // end
+        llvm_utils->start_new_block(loopend);
+
+        return new_list;
+    }
```

Comment:
Logic: I create a new list, deep copy left list, and append all the elements of the right to the new list.

Reply:
Following approaches exist,

1. ASR pass - List concatenation can be easily implemented using Python's syntax. So we can use ASR APIs to replace ListConcat with a temporary result variable. This result variable can be filled using two do loops one which iterates over first list and the other which iterates over second list. Can be easily optimised by tracking if we need a temporary variable or the result can go into some variable on the left hand side of an assignment statement. Pros - Can be shared across all backends at once. Cons - Re-filling the result list using append but not that big of a deal because as the list grows time complexity averages out. For small lists this is not a problem at all.
2. Implementing ListConcat in each backend as you have done for LLVM - We will be able to use `memcpy` and hence this approach might be faster. But that will only work for simple lists of primitive types. For types like `list[str]`, `list[tuple[...]]`, benefits cannot be availed.

I am leaning towards ASR pass approach because our `append` is already fast for large lists, so benefits of second approach should be marginal.

Reply:
I see, then let's work on `ASR pass` for `List Concat`.
I didn't understand properly, how to implement the `ASR pass`?

Reply:
You meant is that to create a function is Python syntax, something like:
```
def list(left: list[i32], right: list[i32]) -> list[i32]:
    l: list[i32] = []
    i: i32
    for i in left:
        l.append(i)
    for i in right:
        l.append(i)
    return l
```
and link this to LLVM backend?

Reply:
I think @czgdp1807 means to create ASR->ASR pass that takes ListConcat ASR node and replaces it with a function call `_lcompilers_list_concat` which implements the list concatenation in Python itself. Then all backends will automatically work with this.

One problem is that we need to figure out a way to make this pass work with LFortran also. But I thin we should figure it out, so that we can share more functions in ASR across compilers and so that we can write more of such passes.

Reply:
Got it, thank you!

Reply:
> I think @czgdp1807 means to create ASR->ASR pass that takes ListConcat ASR node and replaces it with a function call `_lcompilers_list_concat` which implements the list concatenation in Python itself. Then all backends will automatically work with this.

And for that we need generics. We already have those, right? Other way is to do pragmatically using ASR APIs, then you already know what type the lists are and hence generate ASR in the ASR->ASR pass directly.

Reply:
> One problem is that we need to figure out a way to make this pass work with LFortran also.

LFortran has lists? I am not sure. If Fortran doesn't has lists or anything similar then this pass won't have any effect on Fortran code.

Reply:
People keep asking for lists, so we might provide it as an extension.

Also I was thinking providing an optimization that recognizes array usage that would be faster with lists (appending to the end and reallocating), and replace it with a list.

All ASR passes must work with all compilers.

Equivalently: nothing in libasr should be dependent on any one frontend.

We need to design things to work with all frontends.

Reply:
> I think @czgdp1807 means to create ASR->ASR pass that takes ListConcat ASR node and replaces it with a function call `_lcompilers_list_concat` which implements the list concatenation in Python itself. Then all backends will automatically work with this.

I will generate the code for `_lcompilers_list_concat` using ASR APIs, the way we do for vectorisation optimisation. We faced a similar problem with optimisation implemented in LFortran (I listed multiple approaches in https://github.com/lcompilers/lpython/issues/723#issuecomment-1172909436) and we agreed to use ASR APIs (I commented our agreed solution in https://github.com/lcompilers/lpython/issues/723#issuecomment-1181815465) to generate the fallback code so that they can be shared across all the LCompilers. Its the same case here in this PR so I will use the same approach we agreed upon in the past. Do you still want to follow the idea in https://github.com/lcompilers/lpython/issues/723#issuecomment-1181815465 or is there any other approach that can be better?

Reply:
I think that approach still works. I opened up a dedicated issue for this: https://github.com/lcompilers/lpython/issues/1150

Reply:
Cool. Let's try it out here as an experiment (it will work most probably but we will gain more experience if we start earlier). @Thirumalai-Shaktivel Would you like to try it out here? Feel free to let me know if you get stuck anywhere. For replacing expressions you might need to follow the pattern in. Replacing function calls is same as replacing ListConcat as both expression nodes.

https://github.com/lcompilers/lpython/blob/91cb772ef86a58d8b5020a8dde864754bc06d06a/src/libasr/pass/pass_array_by_data.cpp#L303

Reply:
Sure, I will give it a try tomorrow.

Reply:
Sorry, I was working on the parser issue (match statement). I start working tomorrow and report back.

Reply:
No worries. Its not a critical feature IMO, so take as much time as you need to come up with a nice solution.